compiler/rustc_passes/src/liveness.rs

```diff
@@ -329,6 +329,40 @@ impl<'tcx> Visitor<'tcx> for IrMaps<'tcx> {
             }
         }
 
+        // Allow todo! macro
+        /*
+            Skips checking for unused variables when the trailing expression
+            of the body is core::panicking::panic("not yet implemented"), ie
+            as emitted by todo!().
+
+            # Example
+
+            fn foo(x: i32) {
+                // arbitrary code
+                todo!()
+            }
+        */
+        if let ExprKind::Block(block, _) = &body.value.kind {
+            if let Some(expr) = block.expr {
+                if let ExprKind::Call(call, [arg]) = expr.kind {
+                    if let ExprKind::Path(QPath::Resolved(_, path)) = call.kind {
+                        if let Res::Def(DefKind::Fn, path_def_id) = &path.res {
+                            let begin_panic_def_id = self.tcx.lang_items().begin_panic_fn();
+                            if begin_panic_def_id == Some(*path_def_id) {
```

Comment:
The UI test is failing; I think this is the wrong lang item. Sorry about that! I think we need to change this to catch `core::panicking::panic` and `core::panicking::panic_fmt`, but unfortunately there doesn't appear to be a lang item for `panic_fmt`. So you may need to go back to using string comparisons :/

However, I'm out of my depth here so maybe someone else knows a better way (e.g. **`@m-ou-se`**).

If you do end up needing to use the string comparison approach, you may want to use the [`TyCtxt::def_path_str()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.def_path_str) method to get a path string for `path_def_id`. You would use it like:

```rust
let path_str = self.tcx.def_path_str(path_def_id);
```